core/drivers/crypto/se050/adaptors/apis/sss.c

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <config.h>
+#include <crypto/crypto.h>
+#include <se050.h>
+#include <se050_utils.h>
+#include <string.h>
+
+static const sss_policy_u asym_key = {
+	.type = KPolicy_Asym_Key,
+	.auth_obj_id = 0,
+	.policy = {
+		.asymmkey = {
+			.can_Sign = 1,
+			.can_Verify = 1,
+			.can_Encrypt = 1,
+			.can_Decrypt = 1,
+			.can_KD = 1,
+			.can_Wrap = 1,
+			.can_Write = 1,
+			.can_Gen = 1,
+			.can_Import_Export = 1,
+			.can_KA = 1,
+			.can_Read = 1,
+			.can_Attest = 1,
+		}
+	}
+};
+
+static const sss_policy_u common = {
+	.type = KPolicy_Common,
+	.auth_obj_id = 0,
+	.policy = {
+		.common = {
+			.can_Delete = 1,
+			.req_Sm = 1,
+		},
+	},
+};
+
+sss_policy_t se050_asym_policy = {
+	.nPolicies = 2,
+	.policies = { &asym_key, &common },
+};
+
+sss_status_t se050_rotate_scp03_keys(struct sss_se05x_ctx *ctx)
+{
+	struct s050_scp_rotate_cmd cmd = { 0 };
+	sss_status_t status = kStatus_SSS_Fail;
+	struct se050_scp_key cur_keys = { 0 };
+	struct se050_scp_key new_keys = { 0 };
+	SE_Connect_Ctx_t *connect_ctx = NULL;
+	sss_se05x_session_t *session = NULL;
+
+	if (!ctx)
+		return kStatus_SSS_Fail;
+
+	if (IS_ENABLED(CFG_CORE_SE05X_SCP03_EARLY))
```

Comment:
I suggest adding an inline comment about why this fails.
I guess this function should even not be called in such a configuration.

Reply:
it does has to be called and it is supported. but this is a dangerous operation that requires a working RPMB (which I havent tested, only written against secure storage). I am ok to remove the check but I suspect it will lead to a lot of bricked devices if the new random keys generated by this function are not stored in rpmb.

Reply:
You mean se050 mandates `CFG_CORE_SE05X_SCP03_EARLY=y`?
Maybe you should clarify the new `CFG_`s. See comment in **se050/crypto.mk**.

Reply:
sorry @etienne-lms, github drives me nuts (sometimes doesnt allow me to reply to a thread on the right place)..it is pretty annoying.. ....I answered to this question below but will repost here now that I have the option

```
well it is not mandated by an specification but more a feature that we must have: let me recap: SCP03 encrypts the i2c communication with the device and it can be enabled early (ie before the REE is up) or later via PTA.

Doing it early can be done without any risks if we use the pre-programmed keys available in the device (those accessible via the OFID).

But enabling SCP03 early (before REE) with non-default keys requires that we have a persistent place from where to access them (RPMB).
So if we decide to rotate those keys early and then we fail to store them in RPMB the device will be bricked.

Hence why the early SCP enablement can be done without much thinking for default keys but for rotated keys, it requires some thought and care before doing it.
```